soccer_vision_3d_msgs/scripts/visualizer.py

```diff
@@ -0,0 +1,137 @@
+
+    def __init__(self):
+        super().__init__('SVM2DRviz')
+
+        # Create publishers
+        self.marker_publisher = self.create_publisher(Marker, "soccer_vision_3d_msgs", 1)
+
+        # Settings
+        self.ball_diameter = 0.13
+        self.ball_lifetime = int(5e9)
+        self.goal_lifetime = int(5e9)
+        self.obstacle_lifetime = int(5e9)
+        self.robot_lifetime = int(5e9)
+
+        # Ball
+        self.marker_ball = Marker()
+        self.marker_ball.ns = "rel_ball"
+        self.marker_ball.id = 0
+        self.marker_ball.type = Marker.SPHERE
+        self.marker_ball.action = Marker.MODIFY
+        self.marker_ball.scale = Vector3(** dict(zip("xyz", [self.ball_diameter] * 3)))
+        self.marker_ball.color = ColorRGBA(r=1.0, a=1.0)
+        self.marker_ball.lifetime = Duration(nanoseconds=self.ball_lifetime).to_msg()
+
+        # Goalpost
+        self.marker_goalpost = Marker()
+        self.marker_goalpost.ns = "rel_goal"
+        self.marker_goalpost.id = 0
+        self.marker_goalpost.type = Marker.CYLINDER
+        self.marker_goalpost.action = Marker.MODIFY
+        self.marker_goalpost.scale = Vector3()
+        self.marker_goalpost.color = ColorRGBA(r=1.0, g=1.0, b=1.0, a=1.0)
+        self.marker_goalpost.lifetime = Duration(nanoseconds=self.goal_lifetime).to_msg()
+
+        # Robot
+        self.marker_robot = Marker()
+        self.marker_robot.lifetime = Duration(nanoseconds=self.robot_lifetime).to_msg()
+        self.marker_robot.ns = "rel_robot"
+        self.marker_robot.action = Marker.MODIFY
+        self.marker_robot.color = ColorRGBA(a=1.0)
+        self.marker_robot.type = Marker.CUBE
+
+        # Field Boundary
+        self.field_boundary = Marker()
+        self.field_boundary.lifetime = Duration(nanoseconds=self.robot_lifetime).to_msg()
+        self.field_boundary.ns = "rel_field_boundary"
+        self.field_boundary.color = ColorRGBA(g=1.0, a=1.0)
+        self.field_boundary.type = Marker.LINE_STRIP
+        self.field_boundary.action = Marker.MODIFY
+        self.field_boundary.lifetime = Duration(nanoseconds=self.goal_lifetime).to_msg()
+        self.field_boundary.scale = Vector3(x=0.02)
+
+        self.create_subscription(BallArray, "balls_relative", self.balls_cb, 10)
+        self.create_subscription(GoalpostArray, "goal_posts_relative", self.goalpost_cb, 10)
+        self.create_subscription(RobotArray, "robots_relative", self.robot_cb, 10)
+        self.create_subscription(FieldBoundary, "field_boundary_relative", self.field_boundary_cb, 10)
+
+    def conf_to_alpha(self, conf: Confidence):
+        if conf.confidence == Confidence.CONFIDENCE_UNKNOWN:
+            return 1.0
+        return conf.confidence
+
+    def balls_cb(self, msg: BallArray):
+        self.marker_ball.header = msg.header
+        for idx, ball in enumerate(msg.balls):
+            self.marker_ball.pose.position = ball.center
+            self.marker_ball.pose.orientation.w = 1.0
```

Comment:
Why is this necessary?

Reply:
The msg is initialized with zero values. But a quaternion needs to have the length != by definition, so it is standard to set the `w=1`. This is still the identity, but it is a defined value instead of an undefined one.

Reply:
One could save some computation and move it to the initialization part at the top.